base/dict.jl

```diff
@@ -391,6 +391,45 @@ function setindex!(h::Dict{K,V}, v0, key::K) where V where K
     return h
 end
 
+function modify!(f, h::Dict{K}, key0) where K
+    key = convert(K, key0)
+    if !isequal(key, key0)
+        throw(ArgumentError("$(limitrepr(key0)) is not a valid key for type $K"))
+    end
+
+    # Ideally, to improve performance for the case that requires
+    # resizing, we should use something like `ht_keyindex` while
+    # keeping computed hash value and then do something like
+    # `ht_keyindex2!` if `f` returns non-`nothing`.
+    idx = ht_keyindex2!(h, key)
+
+    age0 = h.age
+    if idx > 0
+        @inbounds vold = h.vals[idx]
+        vnew = f(Some(vold))
+    else
+        vnew = f(nothing)
+    end
+    if h.age != age0
+        idx = ht_keyindex2!(h, key)
+    end
+
+    if vnew === nothing
+        if idx > 0
+            _delete!(h, idx)
+        end
+    else
+        if idx > 0
+            h.age += 1
+            @inbounds h.keys[idx] = key
```

Comment:
I copied this line `h.keys[idx] = key` from `get!`. But why is this required? (I'm new to `Dict` code.)

Reply:
What’s going on here is `WeakKeyDict` uses `Dict` internally and sets a finalizar to the keys to mutate the `Dict`. Unfortunately this usage has leaked into the implementation of `Dict`.

Any time the GC is called (to allocate) it may call finalizers which might mutate the dictionary. To protect itself it checks `age` whenever an allocation might occur (depends on the dictionary key and element types, etc).

This works in single threaded concurrency - but I have no idea if the implementation is valid under multithreading?

(@vtjnash did I get those details right?)

Reply:
Thanks!  So, is it effectively `GC.@preserve key`?  And `h.keys[idx] = key` is used because of bootstrapping issue or something?  Or maybe it's just cheaper this way?

Reply:
Sorry I didn't actually address your original question (I was thinking of the `h.age += 1` line... and actually now I look at it I'm not sure that is necessary here or not, since I don't see where the GC might run after `ht_keyindex2!` or how changing `h.vals[idx]` should affect the operation `WeakKeyDict`, but I would defer to Jameson on that).

So the `ht_keyindex2!` function prepares a slot where a key (and value) might reside but doesn't actually populate them with anything. A positive token means the slot already exists, so I think you just need to populate the new value (delete this line).

In the `idx < 0` case the `_setindex!` function populates `h.slots[-idx]`, `h.keys[-idx]` and `h.vals[-idx]`.

Reply:
FYI, existing `get!` method I mentioned was this:

https://github.com/JuliaLang/julia/blob/6eebbbe2d205f8116330a77ca5e15f4a356232db/base/dict.jl#L446-L464

`h.keys[index] = key` was added in https://github.com/JuliaLang/julia/commit/ff4706b1886becb374963f979a35c5187935c053 (#9595)

Reply:
Hmm interesting. Still not sure what it protects against... it would be ideal to hear from Jeff or Jameson, but leaving it in doesn’t hurt (except maybe performance slightly).